core/shared/src/main/scala/scalaz/zio/Queue.scala

```diff
@@ -2,18 +2,66 @@
 
 package scalaz.zio
 
-import scala.collection.immutable.{ Queue => IQueue }
-import Queue.internal._
+import scala.annotation.tailrec
+import scalaz.zio.Queue.internal._
+import scalaz.zio.internal.MutableConcurrentQueue
 
 /**
  *  A `Queue[A]` is a lightweight, asynchronous queue for values of type `A`.
  */
 class Queue[A] private (
-  capacity: Option[Int],
-  ref: Ref[State[A]],
-  strategy: SurplusStrategy
+  queue: MutableConcurrentQueue[A],
+  takers: MutableConcurrentQueue[Promise[Nothing, A]],
+  shutdownHook: Ref[Option[IO[Nothing, Unit]]],
+  strategy: Strategy[A]
 ) extends Serializable {
 
+  private final val checkShutdownState: IO[Nothing, Unit] =
+    shutdownHook.get.flatMap(_.fold[IO[Nothing, Unit]](IO.interrupt)(_ => IO.unit))
+
+  private final val pollTakersThenQueue: IO[Nothing, Option[(Promise[Nothing, A], A)]] = IO.sync {
+    // check if there is both a taker and an item in the queue, starting by the taker
+    val nullTaker = null.asInstanceOf[Promise[Nothing, A]]
+    val taker     = takers.poll(nullTaker)
+    if (taker == nullTaker) {
+      None
+    } else {
+      queue.poll(null.asInstanceOf[A]) match {
+        case null =>
+          takers.offer(taker)
+          None
+        case a => Some((taker, a))
+      }
+    }
+  }
+
+  private final val pollQueueThenTakers: IO[Nothing, Option[(Promise[Nothing, A], A)]] = IO.sync {
+    // check if there is both a taker and an item in the queue, starting by the queue
+    queue.poll(null.asInstanceOf[A]) match {
+      case null => None
+      case a =>
+        val nullTaker = null.asInstanceOf[Promise[Nothing, A]]
+        val taker     = takers.poll(nullTaker)
+        if (taker == nullTaker) {
+          queue.offer(a)
```

Comment:
Do we promise FIFO semantics for the queue? It seems to be possible for items from the head to get moved to the tail of the queue in the following scenario (time moves from top to bottom, interleaving of operations from 2 threads + contents of the queue and takers):
```
| Thread 1                               | Thread 2                  | Queue  | Takers |
|----------------------------------------+---------------------------+--------+--------|
| Queue.take --                          |                           | []     | -      |
| IO.sync(takers.offer(p)) *> ...        |                           |        | p1     |
|                                        | Queue.offerAll([1, 2, 3]) | [2, 3] | -      |
| completeTakers(pollQueueThenTakers) -- |                           |        |        |
| queue.poll == 2                        |                           | [3]    | -      |
| takers.poll == null                    |                           | [3]    | -      |
| queue.offer(2)                         |                           | [3, 2] | -      |
```

If I got something wrong in the algorithm, please let me know.

Reply:
You're right! It's a very unlikely situation where offerAll is processed in between those 2 steps, but it could happen. I will instead poll then re-offer the whole queue instead so that the order is preserved. It won't affect overall performance because 1) it's very unlikely to happen and 2) it only happens when the queue is empty in the first `take` so the queue's size is likely to be small.

Reply:
While fixing a race condition I was actually able to delete this function =)